### PR TITLE
Benchmarks

### PR DESCRIPTION
**Pull request by codebien**
```
goos: linux
goarch: amd64
pkg: github.com/grafana/xk6-output-influxdb/pkg/influxdb
cpu: AMD Ryzen 7 4800H with Radeon Graphics
BenchmarkWritePoints-16    	     192	   6670961 ns/op
PASS
ok  	github.com/grafana/xk6-output-influxdb/pkg/influxdb	1.930s
```

with v1 I got ~220 iteration

### CODE REVIEW

File: pkg/influxdb/bench_test.go [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+		}
+	}
+	batch := o.batchFromSamples([]stats.SampleContainer{samples})
+
+	b.ResetTimer()
+	for i := 0; i < b.N; i++ {
+		err := o.pointWriter.WritePoint(ctx, batch...)
+		if err != nil {
+			b.Fatal(err)
+		}
```
Comment (MStoykov) [inline pkg/influxdb/bench_test.go:158]: I would argue you should be adding Metics the same way the engine would be - through the `AddMetricSamples`. This also will test all of the other parts (like `batchFromSamples`) of writing metrics and arguably can be of help to making something like https://github.com/grafana/k6/issues/2208 more doable.

**Review by MStoykov — COMMENTED** [review]
I am really surprised the influxdb v2 is worse than v1 :( I don't know what the marketed benefits of v2 were supposed to be, but not making ingestion faster seems like ..., not a good thing.

I would argue you are currently benchmarking the `v2` influxdb API, not this extension, to the later I would be calling `AddMetricsSamples` of the extension. This though also needs to take into account the fact that this needs to be flushed. Which is one of the many reasons I do think this things should be tested by just running the whole k6 with just a bunch of custom metrics being added each iteration and waiting for k6 to finish while measuring how many iterations (metric emissions) it was actually possible to send with the given output. 

This though is kind of ... bad for a go benchmark IME, so maybe this is fine as a benchmark :shrug: , it just isn't really showing us how influxdb v2 output compares to v1 or let's say stated with tags :shrug: which arguably is both the thing users will want to know and we can compare and make statements such as "influxdb v2 "is faster/slower than `influxdb v1` output for this script". 

Again my biggest reason to want a direct comparison of script running is that is the thing we actually care about and currently the influxdb output in k6, doesn't ... do well :(. So if this new one does ... worse when it's synchronously writing, maybe it will do better if it does it asynchronously ... or maybe not, but maybe if we change something else ... and so on. But for this we need  to see how it goes from one end to the other.

I would argue also that the major problem of outputs currently is that k6 doesn't aggregate anything and most of them don't either so they end up writing all 20000 updates to a counter every second as 20000 different samples, while arguably no user will care for a resolution under 1s and if they did we can have it be configurable. Once this is fixed I would expect most outputs to be a lot more ... performant ;)

**Comment by codebien** [comment]
>Which is one of the many reasons I do think this things should be tested by just running the whole k6 with just a bunch of custom metrics being added each iteration and waiting for k6 to finish while measuring how many iterations (metric emissions) it was actually possible to send with the given output.

I agree and that is exactly the attempt with `TestOutputThroughput`. I did it in Go just for two reasons:
* it's reproducible
* we can see what is the achieved performance just by looking into the asserted value

Can we achieve the same using the` k6 run --out=...` command? I guess we can have a `bench.js` file in the repo and then report in a section in the README the latest seen value with running the `master` branch?

**Comment by MStoykov** [comment]
Yeah the idea is that we will likely make a script :
* constant arrival rate for consistency and the ability to see if we drop iterations
* maybe constant vus as well to put pressure on it, but in a different script/run :shrug:   
* I would probably just go with custom metrics of each time and just add samples to each of them on each iteration 
* then we look at how long it took to finish and how many iterations were actually finished, maybe memory and CPU usage for the constant arrival rate and w/e else you think of.

p.s. those were three reasons ;)

**Comment by yorugac** [comment]
> I would argue also that the major problem of outputs currently is that k6 doesn't aggregate anything

Totally agree with this, from my observations from PRW output behavior.

A suggestion: it might also be possible to make an estimate of a "metrics rate" as the number of metric samples that are processed by the Output in each flush period. That would allow to re-formulate the problem of output's performance in terms of the metrics rate this given Output can process without errors / dropping samples / etc. For example, the end result would be something like "Output A can handle X samples per second with standard setup™ + addition of 1 custom metric raises the sample rate by Y".
(The last part about custom metrics is likely independent of the Output in question.)

**Comment by codebien** [comment]
At the moment we don't have the time for focusing on this in short term, so I'm closing this and we will reopen maybe a new one better defined in the future. Feel free to reopen or create an issue if you have different opinions.

**Pull request closed without merging**